(window.webpackJsonp=window.webpackJsonp||[]).push([[114],{407:function(t,r,e){"use strict";e.r(r);var a=e(4),n=Object(a.a)({},(function(){var t=this,r=t._self._c;return r("ContentSlotsDistributor",{attrs:{"slot-key":t.$parent.slotKey}},[r("h1",{attrs:{id:"说说你在react项目是如何捕获错误的"}},[r("a",{staticClass:"header-anchor",attrs:{href:"#说说你在react项目是如何捕获错误的"}},[t._v("#")]),t._v(" 说说你在React项目是如何捕获错误的？")]),t._v(" "),r("p",[r("img",{attrs:{src:"https://mmbiz.qpic.cn/mmbiz_png/gH31uF9VIibTib4PFZ0TNDPrPMoFRibY64iblTvt5AcKbpVt0asdxXyl4aXVSvoWWYu5R3bS2oViafmPjhiakqJ4zXrg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1",alt:"图片"}})]),t._v(" "),r("h2",{attrs:{id:"一、是什么"}},[r("a",{staticClass:"header-anchor",attrs:{href:"#一、是什么"}},[t._v("#")]),t._v(" 一、是什么")]),t._v(" "),r("p",[t._v("错误在我们日常编写代码是非常常见的")]),t._v(" "),r("p",[t._v("举个例子，在"),r("code",[t._v("react")]),t._v("项目中去编写组件内"),r("code",[t._v("JavaScript")]),t._v("代码错误会导致 "),r("code",[t._v("React")]),t._v(" 的内部状态被破坏，导致整个应用崩溃，这是不应该出现的现象")]),t._v(" "),r("p",[t._v("作为一个框架，"),r("code",[t._v("react")]),t._v("也有自身对于错误的处理的解决方案")]),t._v(" "),r("h2",{attrs:{id:"二、如何做"}},[r("a",{staticClass:"header-anchor",attrs:{href:"#二、如何做"}},[t._v("#")]),t._v(" 二、如何做")]),t._v(" "),r("p",[t._v("为了解决出现的错误导致整个应用崩溃的问题，"),r("code",[t._v("react16")]),t._v("引用了「错误边界」新的概念")]),t._v(" "),r("p",[t._v("错误边界是一种 "),r("code",[t._v("React")]),t._v(" 组件，这种组件可以捕获发生在其子组件树任何位置的 "),r("code",[t._v("JavaScript")]),t._v(" 错误，并打印这些错误，同时展示降级 "),r("code",[t._v("UI")]),t._v("，而并不会渲染那些发生崩溃的子组件树")]),t._v(" "),r("p",[t._v("错误边界在渲染期间、生命周期方法和整个组件树的构造函数中捕获错误")]),t._v(" "),r("p",[t._v("形成错误边界组件的两个条件：")]),t._v(" "),r("ul",[r("li",[t._v("使用了  static getDerivedStateFromError()")]),t._v(" "),r("li",[t._v("使用了 componentDidCatch()")])]),t._v(" "),r("p",[t._v("抛出错误后，请使用 "),r("code",[t._v("static getDerivedStateFromError()")]),t._v(" 渲染备用 UI ，使用 "),r("code",[t._v("componentDidCatch()")]),t._v(" 打印错误信息，如下：")]),t._v(" "),r("div",{staticClass:"language- extra-class"},[r("pre",{pre:!0,attrs:{class:"language-text"}},[r("code",[t._v("class ErrorBoundary extends React.Component {\n  constructor(props) {\n    super(props);\n    this.state = { hasError: false };\n  }\n\n  static getDerivedStateFromError(error) {\n    // 更新 state 使下一次渲染能够显示降级后的 UI\n    return { hasError: true };\n  }\n\n  componentDidCatch(error, errorInfo) {\n    // 你同样可以将错误日志上报给服务器\n    logErrorToMyService(error, errorInfo);\n  }\n\n  render() {\n    if (this.state.hasError) {\n      // 你可以自定义降级后的 UI 并渲染\n      return <h1>Something went wrong.</h1>;\n    }\n\n    return this.props.children; \n  }\n}\n")])])]),r("p",[t._v("然后就可以把自身组件的作为错误边界的子组件，如下：")]),t._v(" "),r("div",{staticClass:"language- extra-class"},[r("pre",{pre:!0,attrs:{class:"language-text"}},[r("code",[t._v("<ErrorBoundary>\n  <MyWidget />\n</ErrorBoundary>\n")])])]),r("p",[t._v("下面这些情况无法捕获到异常：")]),t._v(" "),r("ul",[r("li",[t._v("事件处理")]),t._v(" "),r("li",[t._v("异步代码")]),t._v(" "),r("li",[t._v("服务端渲染")]),t._v(" "),r("li",[t._v("自身抛出来的错误")])]),t._v(" "),r("p",[t._v("在"),r("code",[t._v("react 16")]),t._v("版本之后，会把渲染期间发生的所有错误打印到控制台")]),t._v(" "),r("p",[t._v("除了错误信息和 JavaScript 栈外，React 16 还提供了组件栈追踪。现在你可以准确地查看发生在组件树内的错误信息：")]),t._v(" "),r("p",[r("img",{attrs:{src:"data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==",alt:"图片"}})]),t._v(" "),r("p",[t._v("可以看到在错误信息下方文字中存在一个组件栈，便于我们追踪错误")]),t._v(" "),r("p",[t._v("对于错误边界无法捕获的异常，如事件处理过程中发生问题并不会捕获到，是因为其不会在渲染期间触发，并不会导致渲染时候问题")]),t._v(" "),r("p",[t._v("这种情况可以使用"),r("code",[t._v("js")]),t._v("的"),r("code",[t._v("try...catch...")]),t._v("语法，如下：")]),t._v(" "),r("div",{staticClass:"language- extra-class"},[r("pre",{pre:!0,attrs:{class:"language-text"}},[r("code",[t._v("class MyComponent extends React.Component {\n  constructor(props) {\n    super(props);\n    this.state = { error: null };\n    this.handleClick = this.handleClick.bind(this);\n  }\n\n  handleClick() {\n    try {\n      // 执行操作，如有错误则会抛出\n    } catch (error) {\n      this.setState({ error });\n    }\n  }\n\n  render() {\n    if (this.state.error) {\n      return <h1>Caught an error.</h1>\n    }\n    return <button onClick={this.handleClick}>Click Me</button>\n  }\n}\n")])])]),r("p",[t._v("除此之外还可以通过监听"),r("code",[t._v("onerror")]),t._v("事件")]),t._v(" "),r("div",{staticClass:"language- extra-class"},[r("pre",{pre:!0,attrs:{class:"language-text"}},[r("code",[t._v("window.addEventListener('error', function(event) { ... })\n")])])]),r("h2",{attrs:{id:"参考文献"}},[r("a",{staticClass:"header-anchor",attrs:{href:"#参考文献"}},[t._v("#")]),t._v(" 参考文献")]),t._v(" "),r("ul",[r("li",[t._v("https://zh-hans.reactjs.org/docs/error-boundaries.html")])])])}),[],!1,null,null,null);r.default=n.exports}}]);